### DISCUSSION TITLE
Still see old commits in 'Activity' after force push

### DISCUSSION THREAD

**Discussion by gudzilla** [community/community#125351] [Repositories] [answered]
### Select Topic Area

Question

### Body

After I pushed 'bad' commits. I decided to clean up my commit history with `git reset` command. I choose a commit and reset to this commit. And my commit history is now short and clean. Btw my first commit is from main branch.
![Screenshot 2024-05-23 at 15 57 19](https://github.com/community/community/assets/57419893/10f63833-7b3d-4b68-92f5-7d1a9f9fef4e)

 And then I used `git push --force-with-lease`. It forces new commit but the old once are still in the 'Active' tab on github.
<img width="719" alt="Screenshot 2024-05-23 at 15 57 43" src="https://github.com/community/community/assets/57419893/845244ba-49c8-477a-b2fc-0dff6aebf7d8">

Why is it so? Isn't `git push --force` suppose to rewrite remote repository history?

**Comment by arashtabaa** [comment, marked as answer]
When you use git push --force, you're indeed overwriting the history of the branch on the remote repository with your local history. However, GitHub doesn't immediately remove the old commits from the "Active" tab in the repository interface. This is because Git, by design, keeps a reference to old commits for a certain period of time, even after they've been removed from the active branch.

The reason behind this behavior is to provide a safety net for developers. If you accidentally force push and delete important commits, they can still be recovered within a certain time frame, usually 30 days by default, through Git's reflog or other recovery mechanisms. During this time, the old commits will still be visible on GitHub's interface.

If you want to completely remove those old commits from GitHub's interface, you would typically have to wait until they expire from the system, which usually takes around 30 days. After that period, the old commits will no longer be visible in the repository's history on GitHub.

So, while git push --force does rewrite the remote repository's history, the old commits may still be visible on GitHub for a limited period of time as a safety measure. If you need the changes to take effect immediately, you might want to contact GitHub support for assistance, but keep in mind that the process of permanently removing commits from the history can be irreversible and should be done with caution.

**Reply by gudzilla** [reply]
Ok, thanks for a detailed answer.

But what if I merge my current branch with main? 
Even though I can see old commits in 'Activity' tab on github in an actual repository there are no old commits right? And I don't see then on `git log` anymore.
These overwritten commits will not take place anywhere else besides my 'Activity' tab of current branch, right?

**Reply by arashtabaa** [reply]
Right, After you perform a git push --force to rewrite the history of a branch, the old commits are no longer part of the active branch history. Here’s a more detailed explanation:

**Post git push --force Behavior:**

1. Local Repository:

- Your local repository history now only contains the commits that you kept after performing git reset and subsequent commits you made.
- The old commits are not visible in your local git log.

2. Remote Repository:

- The remote branch history now matches your local branch history after the forced push. This means the old commits have been removed from the branch’s visible history on the remote.
- However, GitHub retains references to the old commits temporarily, which is why you see them in the "Activity" tab. This does not mean they are part of the current branch’s history.

3. Merging with Main:

- If you merge your current branch with main, the merge will only include the commits that are present in your branch after the forced push.
- The old commits that were removed via the git reset and git push --force will not be part of the main branch history unless they are explicitly reintroduced.
- Confirming No Old Commits in the Active Repository:

4. Git Log:

- The git log will show you the current history of your branch. If the old commits are not in this log, they are not part of the active branch history.

5. GitHub Interface:

- Commits Tab: This will reflect the new history after the forced push.
- Activity Tab: May temporarily show references to the old commits due to GitHub’s safety measures.

6. Overwritten Commits and Repository State:

- Old Commits Visibility:
- Old commits are retained for a short period for recovery purposes but are not part of the active branch.
- They will not be present in any new branches or merges unless manually reintroduced.

7. Current Branch State:

- Your branch, post-merge with main, will only contain the commits present after the forced push.

**Summary:**

> Old commits are not in the actual repository's active history anymore after a `git push --force`.
> Git log reflects the current state of the branch, showing only the new, cleaned-up history.
> Activity tab on GitHub temporarily shows old commits as a safety measure but does not affect the repository's active state.
> Merging with main will not bring back the old commits as long as they are not in the current branch’s history.
> Best Practices Moving Forward:
> Ensure all team members are aware of the history rewrite to avoid confusion.
> Consider using feature branches for development to isolate changes and prevent the need for force pushes on main branches.
> Use `--force-with-lease` to safeguard against accidental overwrites by other collaborators.

**Reply by gudzilla** [reply]
Thanks a lot for such a detailed answer.
+ Respect )

**Comment by njohnso6** [comment]
Hi. I recently did a git push --force and now most of my recent commits are gone. They are not viewable on the webUI or with `git log`, or `git reflog` even after running `git fetch --all`. What could cause such a thing and where can I find them?

**Reply by am-i-abdullah** [reply]
yes please, some help!!
